### ISSUE TITLE
Describe what happens when media changes

### ISSUE THREAD

**Issue by martinthomson** [w3c/webrtc-pc#305]
Sometimes, a change will cause the media to go outside of the parameters that were negotiated.  We could stop sending at that point, but that would be bad.
1. Scale up or down to meet resolution constraints.
2. Letterbox or pillar to comply with aspect ratio constraints.
3. Reduce bitrate to comply with maxfs or bitrate constraints.
4. Resample to comply with sample rate constraints.

Much of this is already done in practice, but never actually specified.

**Comment by jan-ivar** [comment]
See list discussion.

**Comment by alvestrand** [comment]
See also https://github.com/w3c/mediacapture-main/issues/193

**Comment by fluffy** [comment]
This needs a pull request for a API to control the policy in JSEP

**Comment by stefhak** [comment]
At TPAC @fluffy promised to produce a PR.

**Comment by elagerway** [comment]
@fluffy could you provide insight into progress on this one?

**Comment by fluffy** [comment]
still waiting for comments on my outstanding PR before I start others

**Comment by alvestrand** [comment]
@fluffy, if the above is in reference to #434, you haven't replied to the last comment.
This is a discussion topic for tomorrow's interim.

**Comment by alvestrand** [comment]
@fluffy any progress?

**Comment by fluffy** [comment]
Looks like the other error stuff this was blocked on got merged so I will start on this one

**Comment by pthatcherg** [comment]
I think an implementation can always scale down resolution, reduce framerate (or drop frames), and reduce bitrates, if for nothing else than because of bandwidth limitations.    And I thought we already had consensus that we want a knob for allowing scaling up resolution.  

Does that just leave aspect ratio?  For that, I would suggest the answer is to never crop/letterbox/pilar, and always stop sending.  If an application is that picky about getting the aspect ratio right, then its the application's responsibility to match the input track with the negotiated aspect ratio.   

What else is there?

**Comment by alvestrand** [comment]
Chrome today crops in order to satisfy aspect ratio constrains when opening a camera; I'd have to check the code to be sure, but I also think we crop to satisfy width/height ("take the smallest resolution that's larger than target, then crop") - I'd have to re-check the code to be sure; we may rescale instead.
I remember @jan-ivar saying that Firefox does something else, but I'm not sure enough on what he wrote to reproduce it here.

If that's behavior that's satisfactory under all use cases, we can close this issue once we have the down/up-scaling thing settled; if there are other things we should write in the spec, we need to do more.

I'm not sure how often sources change width/height on their own; it may happen if a camera is opened multiple times (opened in VGA first, second opener opens in HD). I know Chrome has an open bug on this (the corresponding bug where we open HD first and then VGA was fixed by making sure we downscale HD->VGA for the second opener.)
Remote streams can change at any time due to sender's decisions, of course.

**Comment by jan-ivar** [comment]
@martinthomson When you say "negotiated" you mean s/constraints/JSEP envelope/ yes?

@alvestrand On the mediacapture front, Firefox's `getUserMedia` is a pure camera discovery API that neither crops, scales, nor decimates frames, on the logic that "if you get exactly what you want then you've discovered nothing" (That said, we're [feeling the pressure to rescale](https://bugzil.la/1286945), though we have no plans to crop since it's [easily done on display with CSS](https://jsfiddle.net/ut4x94ou/)). But cropping seems orthogonal to this issue since a source may change from a 4:3 native mode to a 16:9 native mode even without this feature.

We do processing in peer connection instead, with `scaleResolutionDownBy` and `maxBitrate` (we don't have `maxFrameRate` yet).

**Comment by stefhak** [comment]
@alvestrand I guess one reason for sudden changes to width/height can be changes to the constraints applied to the track. 

I think we should be a little bit clearer in the spec on the relation between track constraints and JSEP envelope.

#624 goes a bit, but in its current form it talks what the "source camera" can produce, and skips that constraints may be applied.

**Comment by jan-ivar** [comment]
@stefhak Not all tracks have constraints (e.g. from `element.captureStream`, `canvas.captureStream`, `audioContext.createMediaStreamDestination`, `peerConnection.ontrack`), and these can still be sent over a peer connection. so I think the abstraction to not talk about constraints in the webrtc spec is an appropriate separation. That said, I welcome any language that makes this clearer.

**Comment by jan-ivar** [comment]
We must assume dimensions of video tracks from these sources (except for the audio one) may change.

**Comment by stefhak** [comment]
@jan-ivar I have the impression that all tracks can be constrained.

**Comment by jan-ivar** [comment]
@stefhak Successfully? Was that the intent of https://github.com/w3c/mediacapture-main/pull/321 ? If so, then this seems woefully under-specified, as [previously](https://lists.w3.org/Archives/Public/public-media-capture/2014Nov/0002.html) this was not the case, and I've seen no new language introduced in other specs suggesting what should happen. It's also worth mentioning that, unlike `getUserMedia`, none of the other sources I mentioned take constraints as an argument at creation.

Was this done to give browsers the option to add rescalers everywhere? If so, I'd like to point out that A) rescaling at multiple points can lead to quality and performance issues, and B) the ConstrainablePattern is a crummy API for a rescaler anyway, since it lets some browsers implement it and others not, dumping the problem on web developers to deal with browser differences. @burnburn ?

In any case, sources [can change for reasons other than constraints](https://github.com/w3c/mediacapture-main/issues/394) and some sources may not respond to constraints at all.

**Comment by jan-ivar** [comment]
Does the `facingMode` constraint on a remote track tell me which way my peer's camera is pointing? If so, then this all has to be implemented (and specified IMHO, not just inferred), otherwise, no it doesn't.

**Comment by stefhak** [comment]
I think ```facingMode``` of a remote track will return ```null``` (in fact there is a sentence in the WebRTC 1.0 spec saying so).

I have a vivid memory of us discussing that the (API) user _can_ apply constraints to a remote (PeerConnection sourced) track. If the constraints (mandatory) required a higher resolution or framerate (etc.) than what is supplied by RTP there would be an overconstrained event (and the track would mute), but downscaling would work. 

I assumed the same behavior would carry over to canvas, video tag, etc. sources. But I think this is a good discussion to have.

**Comment by stefhak** [comment]
@jan-ivar I had the chance to look into -screen-share, and there is actually text about use of constraints there (e.g. ```...constraints apply to the MediaStreamTrack instances that are returned. This allows for changes such as adjustments to frame rate or resolution, or other constraints.```).

**Comment by jan-ivar** [comment]
Yes, we implement constraints for screen sharing. Screen sharing is a gUM source, so that makes sense. I'm less sure about the others I mention. None of them mention anything, so I think we have to assume they don't support it, until they specify what it would mean.

**Comment by stefhak** [comment]
@jan-ivar They all use mediaStreamTrack, and thus (unless anything else is specced) do have methods like getCanstraints, setConstraints, getSettings, so my assumption would be that they _do_ support it.

**Comment by jan-ivar** [comment]
@stefhak Yes, but a track is an interface to a source, so unless the spec that defines that source mentions explicitly which constraints it supports, there's nothing to implement.

**Comment by fluffy** [comment]
Ignore constraints here for a second. Lets say a browser is sending video on a RTP stream that is negotiated to be exactly 100x100 and the source is a canvas that is 200x200. Pretty clear the default behavior that would be desirable is for it to be scaled down by factor of 2 and send. Now lets say the canvas was resized to 200x100. We have two viable option - scale and insert black bars or crop. We discussed this multiple times and clear WG wants to crop to the 200x100 to an 100x100 image then send. Are we all on same page on this part ?

**Comment by alvestrand** [comment]
That's what I interpreted as the consensus of the TPAC discussion, yes.
The sentiment I heard was that if there is a need for any other option, that would be an extension spec.

**Comment by stefhak** [comment]
I also heard that ("the UA should center, scale and crop (in that order) to make the video fit").
